### PR TITLE
Reduce heap contention in Tokenizer

### PR DESCRIPTION
**Pull request by yuslepukhin**
### Description

Re-use vector buffers to prevent frequent reallocations.

### Motivation and Context

Reduce process heap contention.

image

### CODE REVIEW

**Review by pranavsharma — APPROVED** [review]

**Review by edgchen1 — DISMISSED** [review]

**Comment by yuslepukhin** [comment]
I will need to re-work this.

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 1]
```diff
@@ -160,69 +192,159 @@ Status Tokenizer::CharTokenize(OpKernelContext* ctx, size_t N, size_t C,
...
+  allocated_size = size_bytes;
+  return std::align(alignment, size_bytes, ptr, allocated_size);
+}
+
+/// <summary>
+/// This class provides a thin abstraction over the std::pmr::monotonic_buffer_resource
+/// If the allocated buffer is not enough, additional allocations are done using
+/// new/delete.
+/// </summary>
+class MonitonicAllocatorWithDefault : public std::pmr::monotonic_buffer_resource {
```
Comment (pranavsharma) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc:253]: Monotonic

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 2]
```diff
@@ -114,6 +136,25 @@ Tokenizer::Tokenizer(const OpKernelInfo& info) : OpKernel(info) {
   }
 }
 
+Status Tokenizer::EstimateNumberOfTokens(gsl::span<const std::string> input_span,
+                                         size_t& max_tokens_per_row, size_t& total_tokens_estimate) const {
+  total_tokens_estimate = 0;
+  max_tokens_per_row = 0;
+  for (const auto& s : input_span) {
```
Comment (xadupre) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc:143]: Should you include StartMarker and EndMarker if mark_ is true? Maybe it is added later.

Comment (yuslepukhin) [reply onnxruntime/contrib_ops/cpu/tokenizer.cc:143]: This is done later.

**Review by pranavsharma — DISMISSED** [review]

**Pull request merged**